dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/BitstreamResourceAccessByToken.java

```diff
@@ -0,0 +1,203 @@
+        }
+
+        try (Context fileRetrievalContext = initializeContext()) {
+            // Set special privileges for context for this access
+            // Note - this is a try-with-resources statement. The context has a very limited lifetime.
+            // However, be very careful using context in this block! It should ONLY perform authorization
+            // of the access token and retrieval of the bitstream content.
+            fileRetrievalContext.turnOffAuthorisationSystem();
+            // Get bitstream from uuid
+            Bitstream bitstream = bitstreamService.find(fileRetrievalContext, uuid);
+
+            try {
+                // Explicitly authenticate the access request acceptance for the bitstream
+                // even if we have already done it in the REST controller and throw Authorize exception if not valid
+                requestItemService.authorizeAccessByAccessToken(fileRetrievalContext, bitstream, this.accessToken);
+
+            } catch (AuthorizeException e) {
+                throw new AuthorizeException("Authorization to bitstream " + uuid + " by access token FAILED");
+            }
+
+            // Initialise outstream
+            InputStream out;
+
+            if (shouldGenerateCoverPage) {
+                out = new ByteArrayInputStream(getCoverpageByteArray(fileRetrievalContext, bitstream));
+            } else {
+                out = bitstreamService.retrieve(fileRetrievalContext, bitstream);
+            }
+
+            this.file = null;
+            return out;
+
+            // This is the last line in the closeable Context try-with-resources.
+            // If this scope completes, there should be no more context.
+
+        } catch (SQLException | AuthorizeException e) {
+            throw new IOException(e);
+        }
+    }
+
+    /**
+     * Get the filename of the download disposition
+     * @return file name
+     */
+    @Override
+    public String getFilename() {
+        return name;
+    }
+
+    /**
+     * Get content length of the bitstream
+     * @return content length
+     * @throws IOException
+     */
+    @Override
+    public long contentLength() throws IOException {
+        try (Context context = initializeContext()) {
+            Bitstream bitstream = bitstreamService.find(context, uuid);
+            if (shouldGenerateCoverPage) {
+                return getCoverpageByteArray(context, bitstream).length;
+            } else {
+                return bitstream.getSizeBytes();
+            }
+        } catch (SQLException | AuthorizeException e) {
+            throw new IOException(e);
+        }
+    }
+
+    /**
+     * Initialise a new temporary context just for the use of this download
+     *
+     * @return totally new context
+     * @throws SQLException
+     * @throws AuthorizeException
+     */
+    private Context initializeContext() throws SQLException, AuthorizeException {
+        return new Context();
```

Comment:
We might consider making this a READ-ONLY context.  `new Context(Mode.READ_ONLY)`.